### ISSUE TITLE
CF filtering outpoints vs scriptpubkeys

### ISSUE THREAD

**Issue by cpacia** [gcash/bchd#205]
The original bip157/158 proposal called for the block filter to consist of:

- The scriptPubkey of every output in the block
- The outpoint of every input in the block

This would allow a client that was confronted with two different filter hashes to interrogate the filters by downloading the filter from each peer, downloading the block, calculating the correct filter from the block, and banning the peer which sent it the invalid filter. 

Greg Maxwell basically demanded the `outpoints` be removed from the filter and replaced with the input `scriptPubkeys` because:
- They would be more useful (unclear to me how)
- "I think lite clients cross checking is something which very likely will never be implemented by anyone" (this was already implemented in code at the time he wrote this comment). 
- Eventually the filters should be committed at which point interrogating the filters would not be necessary.

Greg also says:
> Monitoring inputs by scriptPubkey vs input-txid also has a massive
advantage for parallel filtering:  You can usually known your pubkeys
well in advance, but if you have to change what you're watching block
 N+1 for based on the txids that paid you in N you can't filter them
in parallel.

But this seems to be false on the surface. Any wallet with an hdkeychain doesn't know the full set of keys up front. While scanning for transactions, every time you get a match you extend the keychain by one. 

If you scan block ranges `a to b` and `c to d` in parallel `a to b` might contain transactions that require extending the keychain. If you scan `c to d` without those extended keys you would miss your transactions. Seems like you would go from "you will definitely detect all your transactions" to "you will detect all your transactions with some probability based on the size of the lookahead window". 

This has created a big problem in the Neutrino client as a client confronted with two different filter chains basically deadlocks because it can't figure out which chain is correct since it doesn't have the `scriptPubkeys` needed to calculate the filter. 

Here is the relevant discussion on the Bitcoin mailing list https://www.mail-archive.com/bitcoin-dev@lists.linuxfoundation.org/msg07028.html

The big question for us seems to be do we switch the filter format back to `outpoints` from `scriptPubkeys`?

If we don't we are leaving open a denial of service attack vector for the Neutrino wallet. 

If we do this opens up a couple more questions:
- If we ever decide to commit the filters, would we prefer `scriptPubkeys` over `outpoints`?
- If yes, would it be too awkward to switch filter formats at the time of commitment? Doing so would basically require either all clients upgrade prior to activation date so that they can switch filter formats on activation, or old clients would need to continue to connect to a non-upgraded node while upgraded clients connect to upgraded nodes.

**Comment by zquestz** [comment]
@jimpo would love your thoughts on this.

**Comment by markblundeberg** [comment]
A random thought: for committed filters it could still be nice to have a separate filter for outpoints and output-scriptpubkeys; and moreover to modify the golomb code for the latter to be multiset filter (this could be done by allowing delta=0 to be encoded in one bit-- adding one bit per entry, plus one bit per reuse).

Such a multiset allows light wallets to request that a node provides all transactions with outputs matching a given filter entry (compromise privacy, but not needing to download entire block); but, they can check the response to make sure that all matches were correctly provided.

**Comment by jimpo** [comment]
Yeah, I still think outpoints are the better choice, at least until there are either filter commitments or UTXO commitments. Splitting the filters into multiple (output scripts and outpoints/prevout scripts) may make it easier to upgrade later on and the loss in compression efficiency is pretty negligible after the first ~100 elements in the filter.

The multiset idea might work, but I'd be concerned that it loses pretty much all advantages over BIP 37 (both privacy and server-side resource requirements).

Given the size of blocks, you might also want to look into multi-level filter designs. Essentially having multiple higher-FP-rate filters per block element, and if the first level matches, you fetch the next level. You can basically trade rounds of interaction for reduced bandwidth.

**Comment by cpacia** [comment]
Thanks for the feedback. We're going to go back to outpoints. If we ever decide to commit the filter we'll need to revisit the format.

**Comment by markblundeberg** [comment]
@jimpo quick question before you run off -- I saw this in an old review from core https://github.com/bitcoin/bitcoin/pull/12254, which apparently got taken out
![Screenshot from 2019-04-10 13-57-46](https://user-images.githubusercontent.com/36528214/55913229-fed46600-5b98-11e9-9f39-63e5181272d0.png)
I know something similar is being considered in BCH too (https://github.com/bitcoincashorg/bitcoincash.org/pull/311). I worry slightly about it in that someone can be annoying by spamming numerous short and distinct op_return pushes. Given that BCH opreturn relay limit is 220 bytes and unlimited numbers of pushes are allowed, it may take only 6 bytes onchain to produce new filter items, which compress down to only 3 bytes. This could cause some large filters. Is this worth worrying about seriously, or nah? We might for example restrict to only the first push...

**Comment by markblundeberg** [comment]
> Greg also says:
> 
> > Monitoring inputs by scriptPubkey vs input-txid also has a massive
> > advantage for parallel filtering:  You can usually known your pubkeys
> > well in advance, but if you have to change what you're watching block
> > N+1 for based on the txids that paid you in N you can't filter them
> > in parallel.
> 
> But this seems to be false on the surface. Any wallet with an hdkeychain doesn't know the full set of keys up front. While scanning for transactions, every time you get a match you extend the keychain by one.
> 
> If you scan block ranges `a to b` and `c to d` in parallel `a to b` might contain transactions that require extending the keychain. If you scan `c to d` without those extended keys you would miss your transactions. Seems like you would go from "you will definitely detect all your transactions" to "you will detect all your transactions with some probability based on the size of the lookahead window".

I have been thinking about this a bit more, I don't think it becomes lossy if one is careful. You can include a big HD gap limit and process multiple blocks' filters in chunks. If suddenly you hit a filter that looks like it's close to overflow the lookahead window (rare), it's always possible to rescan the subsequent filters with a bigger gap limit. This is the same kind of care one needs to take even when processing block contents (someone could surprisingly overflow the gap limit within one block, which would necessitate a rescan of said block).

In contrast if you're forced to scan for outpoints to find the spends, it makes this kind of chunking not really possible anymore. That said, I don't think parallelization is a big deal at all. Meh.

Another point I'm surprised Greg didn't bring up is that filtering for outpoints will necessarily increase the number of false positives. It's negligible for most wallets where the number of addresses exceeds the number of utxos. But you get crazy donation wallets like the [eatbch one](https://blockchair.com/bitcoin-cash/address/pp8skudq3x5hzw8ew7vzsw8tn4k8wxsqsv0lt0mf3g) which is a single address that I saw at times containing 10000 utxos -- they're now spent on BCH, yet [still largely unspent on BSV at the moment so I took a snapshot](http://archive.is/TaEM3). The false positive rate for the scriptPubKey filter will be 1/784931 since it's a single-address wallet. The rate for the outpoint filter with 10000 utxos will be ~10000/784931 = 1.3%, which means a significant increase in the overall amount of data needing to be downloaded.

Anyway just some things to think about ... for sure when we do commitments I think we ought to do just scriptpubkeys input&output

**Comment by jimpo** [comment]
@markblundeberg The motivation for no OP_RETURN is 1) to reduce filter size, and 2) to make it easier to do a future soft fork commitment in an OP_RETURN output of the coinbase tx (which would create a circular dependency otherwise). I think creating a separate filter for OP_RETURN data is probably fine, though clients then leak that they are an application that cares about OP_RETURN data.
